### ISSUE TITLE
lossの求め方について

### ISSUE THREAD

**Issue by nezura** [miyamotok0105/pytorchapibook#4]
エポックごとのlossを求める際に、データ件数で割り算されています。
例　section3_1.ipynb
avg_train_loss = train_loss / len(train_loader.dataset)
ミニバッチ数で割る場合が多いようですが、違いはありますか？
avg_train_loss = train_loss / len(train_loader)

**Comment by daikiclimate** [comment]
> エポックごとのlossを求める際に、データ件数で割り算されています。
> 例　section3_1.ipynb
> avg_train_loss = train_loss / len(train_loader.dataset)
> ミニバッチ数で割る場合が多いようですが、違いはありますか？
> avg_train_loss = train_loss / len(train_loader)

avg_train_loss自体はprintするためだけに計算されているので、ぶっちゃけ結果には影響ありません
(loss.backward()したあとなので)
なので、違いはあるかというと、表示される数字の桁数が変わるくらいで結果には影響はないと思います。

ただ、このケースで適切なロスは後者の

```
avg_train_loss = train_loss / len(train_loader)
```

と思います。

すでに

```
loss = criterion(outputs, labels)
```

の中で、ミニバッチ内でロスが平均化されているので、データの件数で割るものではないです

**Comment by nezura** [comment]
回答ありがとうございます。
承知いたしました。

**Comment by hei4** [comment]
ご指摘ありがとうございます。ご指摘の通り、加算した損失値をミニバッチの個数（データローダーからロードされるミニバッチの数のこと）で除算するのが適切です。

ご指摘を反映させていただきました。

なお補足説明させていただきますと、損失関数クラスの初期化メソッドにおいて引数reductionで損失値の集約方法を指定することができます。

```
criterion = nn.CrossEntropyLoss(reduction='elementwise_mean') # ミニバッチ内の平均（書籍が対応するPyTorch0.4.1での書き方）
# criterion = nn.CrossEntropyLoss(reduction='mean') # PyTorch1.系での書き方

criterion = nn.CrossEntropyLoss(reduction='sum') # ミニバッチでの損失値の和

criterion = nn.CrossEntropyLoss(reduction='none') # 集約なし
```

引数reductionを指定しない場合、集約方法にはelementwise_meanが設定されます。

損失関数クラスのインスタンスに reduction='elementwise_mean' が設定されていた場合、ご指摘のようにミニバッチの個数で除算するのが適切です。

ただしDataLoaderでdrop_last=Falseにしている場合、毎エポックの最後のミニバッチだけバッチサイズが小さくなる可能性があるため、わずかに偏りが生じる危険がありますので注意してください。

より正確な値を求めたい場合は、損失関数クラスを reduction='sum' に設定し、加算した損失値をデータの個数（データの件数）で除算してください。

ミニバッチの個数が大きく最後のミニバッチの影響が無視できる場合や、簡便に毎エポックを代表する損失値を求めたい場合は、reduction='elementwise_mean' でミニバッチの個数で除算して問題ないと考えます。サンプルコードも後者の考えに従い、そのようにさせていただきました。

改めて、nezuraさんのご指摘に御礼申し上げます。またdaikiclimateさんのフォローに感謝いたします。ありがとうございました。
